### PR TITLE
Regularization and caching in BASEX

### PR DESCRIPTION
**Pull request by MikhailRyazanov**
Enabled Tikhonov regularization in BASEX (see #225) and implemented RAM caching (see #226).
(Sorry, no unit tests.)

### CODE REVIEW

File: abel/basex.py [hunk 1]
```diff
@@ -26,8 +26,12 @@
 # V. Dribinski, A. Ossadtchi, V. A. Mandelshtam, and H. Reisler,
 # Review of Scientific Instruments 73, 2634 (2002).
 #
-# 
-# version 0.62 - 2016-03-07 
+#
+# Version 0.63 - 2018-09-08
```
Comment (DanHickstein) [inline abel/basex.py]: These "version" label are no longer needed. You can remove them and just have the date and the change.

File: abel/basex.py [hunk 2]
```diff
@@ -174,23 +180,23 @@ def basex_core_transform(rawdata, M_vert, M_horz, Mc_vert,
...
-def _get_left_right_matrices(M_vert, M_horz, Mc_vert, Mc_horz):
+def _get_left_right_matrices(M_vert, M_horz, Mc_vert, Mc_horz, reg):
     """ An internal helper function for no-up/down-asymmetry BASEX:
         given basis sets  M_vert, M_horz, Mc_vert, Mc_horz,
         return M_left and M_right matrices
     """
 
     nbf_vert, nbf_horz = np.shape(M_vert)[1], np.shape(M_horz)[1]
-    q_vert, q_horz = 0, 0  # No Tikhonov regularization
+    q_vert, q_horz = reg, reg  # Tikhonov regularization parameters
```
Comment (DanHickstein) [inline abel/basex.py]: Should the regularization be applied to both the horizontal or to both the horizontal and vertical (as done here)? I think that perhaps it should only be applied to the horizontal...

File: abel/basex.py [hunk 3]
```diff
@@ -51,40 +55,42 @@
...
         The image to be inverse transformed. The width (M) must be odd and ``data[:,0]``
-        should correspond to the central column of the image. 
+        should correspond to the central column of the image.
     nbf : str or list
         number of basis functions. If ``nbf='auto'``, it is set to ``(n//2 + 1)``.
         *This is what you should always use*,
         since this BASEX implementation does not work reliably in other situations!
         In the future, you could use
         list, in format [nbf_vert, nbf_horz]
+    reg : float
```
Comment (DanHickstein) [inline abel/basex.py]: This needs a longer description about how this is the tikhonov regularization factor can smooth the data, and provide some estimate of how to select and appropriate value.

File: abel/basex.py [hunk 4]
```diff
@@ -259,54 +273,70 @@ def get_bs_basex_cached(n_vert, n_horz,
         the matrices that compose the basis set.
     """
 
+    global _prm, _M, _reg, _LR
```
Comment (DanHickstein) [inline abel/basex.py]: Yikes, is making these global variables the only way to cache these things?

**Comment by DanHickstein** [comment]
Hi @MikhailRyazanov, thank you so much for doing this!

I made a few specific comments in the code, but overall I think that your changes look really good.

Your new code (with reg=0) should exactly reproduce the old basex behavior exactly, correct? If so, it would be nice to confirm this behavior.

I think that my biggest questions are:

1) Is there an alternative to using global variables to store the basis sets? We don't need these variables to be truly global, they only need to exist in the `abel.` namespace. This might take me some googling and learning more about Python. 

2) Should the regularization be applied to horizontal+vertical, or just horizontal. Or, should we allow users to specify the `(horiz, vert)` tuple of Tikhonov regularization factors?

File: abel/basex.py (hunk 4 above)
Comment (MikhailRyazanov) [reply abel/basex.py]: They are local to this module. From user's perspective, they are in principle accessible as `abel.basex._prm` and so on, but otherwise have no effect.

Comment (DanHickstein) [reply abel/basex.py]: Okay, that seems reasonable.

**Comment by DanHickstein** [comment]
So, I think that we are converging towards merging this. I think that the biggest lingering issue in my mind is if the regularization should be applied in the vertical direction or just in the horizontal.

**Comment by MikhailRyazanov** [comment]
> Your new code (with reg=0) should exactly reproduce the old basex behavior exactly, correct? If so, it would be nice to confirm this behavior.

I've compared them using test images, and yes, they give bitwise equal results.

The old code did
```
q_vert, q_horz = 0, 0
```
and the new does
```
reg=0.0
q_vert, q_horz = reg, reg
```
which has exactly the same effect on the rest of the code.

**Comment by MikhailRyazanov** [comment]
> Still, an open question about the vertical regularization factor, right?

It should not be applied, see https://github.com/PyAbel/PyAbel/issues/225#issuecomment-421698132.

**Comment by stggh** [comment]
Fantastic work @MikhailRyazanov! You have transformed `basex` into one of the fastest (and most accurate) methods (when using a precomputed basis).

For the benchmark, beyond the first calculation, `basex` will use its cached basis (this may be a *feature*). Set `abel.basex._prm = None` on line ~120, before the basis calculation call in `benchmark.py`?
```
In [2]: abel.benchmark.AbelTiming(n=[501], select=['basex', 'hansenlaw', 'three_
   ...: point'])
Out[2]: 
PyAbel benchmark run on x86_64

time in milliseconds

=========    Basis Abel implementations ==========

Implementation      n = 501       
----------------------------------------
        basex_bs   17300.4         
  three_point_bs       8.4         


=========  Forward Abel implementations ==========

Implementation      n = 501       
----------------------------------------
       hansenlaw      27.0         


=========  Inverse Abel implementations ==========

Implementation      n = 501       
----------------------------------------
           basex       1.0         
       hansenlaw      27.1         
     three_point       0.7         

In [3]: abel.benchmark.AbelTiming(n=[501], select=['basex', 'hansenlaw', 'three_
   ...: point'])
Out[3]: 
PyAbel benchmark run on x86_64

time in milliseconds

=========    Basis Abel implementations ==========

Implementation      n = 501       
----------------------------------------
        basex_bs       0.0         
  three_point_bs       8.8         


=========  Forward Abel implementations ==========

Implementation      n = 501       
----------------------------------------
       hansenlaw      27.5         


=========  Inverse Abel implementations ==========

Implementation      n = 501       
----------------------------------------
           basex       1.0         
       hansenlaw      25.8         
     three_point       0.8         
```

**Comment by MikhailRyazanov** [comment]
> For the benchmark, beyond the first calculation, basex will use its cached basis (this may be a feature). Set abel.basex._prm = None on line ~120, before the basis calculation call in benchmark.py?

Isn't this basis generation called only once?

Regarding transforms, I was also thinking about adding tests with non-zero `reg` and with variable `reg` (updating regularization every time). Maybe non-cached test can be added as well.

> Minor `pycodestyle` corrections:

Thanks! I'll go through them later. Please also check whatever else possible and report here.

**Comment by stggh** [comment]
> Isn't this basis generation called only once?

In general yes, but a second call e.g. adding/comparing another method, reports the cached basis fetch time for `basex`:
```python
abel.benchmark.AbelTiming(n=[501], select=['basex'])
abel.benchmark.AbelTiming(n=[501], select=['basex', 'three_point'])
```
(output snipped)

File: abel/basex.py [hunk 5]
```diff
@@ -134,7 +134,7 @@ def basex_transform(data, nbf='auto', reg=0.0, basis_dir='./', dr=1.0, verbose=T
     recon = basex_core_transform(data, Ai, dr)
 
     if data_ndim == 1:  # taking the middle row, since the rest are zeroes
-        recon = recon[recon.shape[0] - recon.shape[0]//2 - 1] #??
+        recon = recon[recon.shape[0] - recon.shape[0]//2 - 1]  # ??
```
Comment (MikhailRyazanov) [inline abel/basex.py]: What is this supposed to do?
`data_ndim == 1` means that recon.shape[0] would be 1, and there are no other rows. Should this be just `recon = recon[0]`?

**Comment by MikhailRyazanov** [comment]
> but a second call e.g. adding/comparing another method, reports the cached basis

Ah, I see. Let me do that after implementing the clean-up procedure. Then it could be done in a better way, without directly poking internal variables.

**Comment by stggh** [comment]
Here are some PyAbel benchmark comparisons, on a i7-6700 (Linux), Anaconda python3.7. `basex` is now ~20-30x faster and now mirrors the Dasch methods execution time (for larger images).

old vs new `basex`:
orig-benchmarks-i7 benchmarks-i7

old basex:
```
#system information: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz4.15.0-34-genericx86_64Linux3.7.0
n,basex,basex_bs,direct_C,direct_C_f,direct_Python,direct_Python_f,hansenlaw,hansenlaw_f,linbasex,linbasex_bs,onion_bordas,onion_peeling,onion_peeling_bs,two_point,two_point_bs,three_point,three_point_bs
5,       0.015,       2.786,       0.135,       0.096,       0.378,       0.375,       0.221,       0.173,       0.447,       0.176,       0.098,       0.022,       0.117,       0.017,       0.071,       0.016,       0.494
9,       0.013,       0.624,       0.160,       0.098,       0.523,       0.464,       0.224,       0.202,       0.382,       0.141,       0.144,       0.022,       0.103,       0.017,       0.118,       0.016,       0.422
17,       0.013,       1.902,       0.144,       0.100,       0.741,       0.730,       0.297,       0.273,       0.471,       0.156,       0.321,       0.023,       0.108,       0.018,       0.079,       0.018,       0.417
33,       0.030,       8.204,       0.174,       0.128,       1.377,       1.341,       0.462,       0.442,       0.909,       0.217,       0.953,       0.029,       0.138,       0.023,       0.104,       0.018,       0.462
65,       0.081,      58.107,       0.285,       0.198,       2.907,       2.826,       0.855,       0.833,       2.013,       0.431,       3.368,       0.034,       0.186,       0.026,       0.177,       0.026,       0.600
129,       0.374,     560.470,       0.971,       0.811,      10.350,      10.285,       2.845,       2.839,       7.597,       1.360,      21.208,       0.061,       0.316,       0.046,       0.319,       0.044,       1.302
257,       1.511,    3851.913,       3.537,       4.089,      36.115,      41.588,       5.753,       8.106,      22.090,       4.704,      57.766,       0.118,       0.814,       0.088,       0.730,       0.086,       3.015
513,      20.428,   18516.776,      23.379,      21.934,     249.350,     253.481,      18.956,      19.063,      90.041,      16.163,     271.115,       0.535,       2.458,       0.456,       2.261,       0.464,       7.286
1025,     131.664,  181369.838,     167.001,     162.361,    1991.546,    2052.152,      77.038,      75.499,     429.933,      68.479,    1314.595,       3.607,       9.098,       3.423,       8.874,       3.407,      28.863
2049,     751.453, 1695030.861,    1315.215,    1245.970,         nan,         nan,     420.008,     434.809,    3598.944,     242.598,   17422.472,      40.300,      39.395,      25.649,      39.363,      25.998,     127.417
```
new basex:
```
#system information: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz4.15.0-34-genericx86_64Linux3.7.0
n,basex,basex_bs,direct_C,direct_C_f,direct_Python,direct_Python_f,hansenlaw,hansenlaw_f,linbasex,linbasex_bs,onion_bordas,onion_peeling,onion_peeling_bs,two_point,two_point_bs,three_point,three_point_bs
5,       0.010,       2.563,       0.131,       0.095,       0.376,       0.386,       0.193,       0.170,       0.451,       0.187,       0.094,       0.022,       0.127,       0.018,       0.072,       0.016,       0.437
9,       0.007,       0.517,       0.147,       0.092,       0.506,       0.470,       0.270,       0.195,       0.390,       0.138,       0.144,       0.024,       0.106,       0.017,       0.073,       0.016,       0.395
17,       0.006,       1.640,       0.143,       0.097,       0.734,       0.693,       0.297,       0.329,       0.479,       0.163,       0.312,       0.022,       0.111,       0.018,       0.075,       0.018,       0.410
33,       0.010,       7.582,       0.176,       0.123,       1.325,       1.263,       0.451,       0.428,       1.068,       0.189,       0.950,       0.025,       0.128,       0.020,       0.097,       0.018,       0.436
65,       0.013,      57.480,       0.285,       0.195,       2.869,       2.843,       0.857,       0.834,       2.027,       0.401,       3.317,       0.038,       0.176,       0.026,       0.135,       0.026,       0.591
129,       0.041,     565.035,       0.877,       0.604,       8.157,       7.833,       2.032,       2.414,       6.392,       1.210,      14.461,       0.065,       0.277,       0.045,       0.249,       0.046,       0.957
257,       0.192,    3849.960,       3.547,       4.067,      37.729,      41.212,       5.758,       7.951,      22.043,       5.318,      59.408,       0.121,       0.831,       0.086,       0.707,       0.086,       2.934
513,       0.647,   18564.039,      22.782,      21.864,     250.937,     258.295,      19.052,      21.462,      90.929,      16.951,     264.744,       0.522,       2.615,       0.475,       2.283,       0.458,       7.357
1025,       5.975,  183628.115,     168.546,     163.259,    1988.648,    2102.666,      77.351,      76.149,     427.017,      66.633,    1314.444,       3.628,       9.264,       3.411,       9.312,       3.413,      28.630
2049,      28.499, 1692925.185,    1308.304,    1292.448,         nan,         nan,     424.429,     428.612,    3799.722,     252.686,   18540.874,      26.081,      42.372,      25.938,      39.966,      26.103,     131.356
```

**Comment by MikhailRyazanov** [comment]
Excellent! Now it can also do real-time HD video.

Could you also try it with variable regularization parameter (say, ramping within 0–1000 or randomly in the same range), such that the transform is updated for each image? I suspect it to be ~5 times slower (for 1 quadrant, ~3 times for a half-image), but still faster than the old one and many other methods.

**Comment by stggh** [comment]
AFAICT the `reg` parameter makes little difference to the execution time.
```
python O2.py 
Loading basis sets...
Updating regularization...
reg = 0 in 123.5 ms
Using memory-cached basis sets
Updating regularization...
reg = 100 in 144.7 ms
Using memory-cached basis sets
Updating regularization...
reg = 1000 in 143.0 ms
```
reg

BTW, `reg=1000` produces better intensity agreement cf `three_point`. Perhaps, the `MAGIC_NUMBER` factor changes with `reg`?

crude code:
```python
t0 = time.time()
basex0 = abel.basex.basex_transform(q[0])
t1 = time.time()
print(f'reg = 0 in {(t1-t0)*1000:.1f} ms')
t0 = time.time()
basex100 = abel.basex.basex_transform(q[0], reg=100)
t1 = time.time()
print(f'reg = 100 in {(t1-t0)*1000:.1f} ms')
t0 = time.time()
basex1000 = abel.basex.basex_transform(q[0], reg=1000)
t1 = time.time()
print(f'reg = 1000 in {(t1-t0)*1000:.1f} ms')
```

**Comment by MikhailRyazanov** [comment]
An illustration of what is mean:

```python
def timing(q, reg=0.0):
    t0 = time.time()
    rec = abel.basex.basex_transform(q, reg=reg)
    print('reg = {} in {:.1f} ms'.format(reg, (time.time() - t0) * 1000))

timing(q)        # (1)
# Loading basis sets...
# Updating regularization...
# reg = 0.0 in 456.2 ms

timing(q)        # (2)
# Using memory-cached basis sets
# reg = 0.0 in 83.4 ms

timing(q, 100)   # (3)
# Using memory-cached basis sets
# Updating regularization...
# reg = 100 in 481.4 ms

timing(q, 100)   # (4)
# Using memory-cached basis sets
# reg = 100 in 84.1 ms

timing(q, 1000)  # (5)
# Using memory-cached basis sets
# Updating regularization...
# reg = 1000 in 483.4 ms

timing(q, 1000)  # (6)
# Using memory-cached basis sets
# reg = 1000 in 85.4 ms
```

(2), (4) and (6) use the cached transform matrix, so they are fast and do not depend on the value of `reg`.
(1), (3) and (5) need to update the transform matrix, so they are slower, ~5–6 times here. (In principle, (1) with `reg=0` should be somewhat faster than (3) and (5), if it was not loading the basis.)

For video processing with constant parameters, we care about timings like (2), (4) and (6). But for interactive optimization of the regularization parameter, timings like (3) and (5) are important. Having separate benchmarks for these two cases would be useful.

**Comment by MikhailRyazanov** [comment]
> BTW, `reg=1000` produces better intensity agreement cf `three_point`. Perhaps, the `MAGIC_NUMBER` factor changes with `reg`?

In short, Tikhonov regularization does not preserve normalization (neither relative intensities). The effect is small for reasonable regularization strengths. And since the transform is still linear, different images with the same regularization are directly comparable. But comparison of images with different regularizations should be done carefully.

Details:
Instead of solving
(1) **Ax** = **y**
it finds
(2) **y** = argmin (||**Ax** − **y**|| + reg·||**x**||).
If reg = 0, they are equivalent (if **A**<sup>−1</sup> exists).
If reg > 0, (2) finds an approximation to (1), preferring least-norm solutions. The large is the reg value, the more it tend to minimize ||**x**|| rather than ||**Ax** − **y**||, and in the limit reg = ∞ we get **x** = **0** regardless of **A**.

What you see in the speed distribution is partially due to that effect and partially due to the fact the `three_point` and `basex reg=1000` peaks are slightly broader (and thus should be slightly lower).

**Comment by DanHickstein** [comment]
Thanks @MikhailRyazanov for all of these great improvements! The BASEX speedup is fantastic, and it makes sense that it would take a similar amount of time to the three-point method, since they all boil down to some fairly similar matrix operations.

> For video processing with constant parameters, we care about timings like (2), (4) and (6). But for interactive optimization of the regularization parameter, timings like (3) and (5) are important. Having separate benchmarks for these two cases would be useful.

Yeah, I suppose that two different benchmarks is the most complete way to describe this situation. If we were really trying to make things seem fast, at the risk of increasing complexity, we could also save the basis sets for each regularization parameter to disk and then lump the calculation into the basis set calculation.

> What you see in the speed distribution is partially due to that effect and partially due to the fact that the three_point and basex reg=1000 peaks are slightly broader (and thus should be slightly lower).

That makes sense to me.

File: abel/basex.py [hunk 6]
```diff
@@ -110,280 +120,289 @@ def basex_transform(data, nbf='auto', basis_dir='./', dr=1.0, verbose=True,
...
 
-    vert_left  = scipy.dot( inv(scipy.dot(Mc_vert.T, Mc_vert) + E_vert),  Mc_vert.T)
-    horz_right = scipy.dot( M_horz, inv(scipy.dot(M_horz.T, M_horz) + E_horz) )
-    
-    # previously: 
-    # vert_left = inv(Mc_vert.T.dot(Mc_vert) + E_vert).dot(Mc_vert.T)
-    # horz_right = M_horz.dot(inv(M_horz.T.dot(M_horz) + E_horz))
-    
+    if reg == 0.0:
+        Ai = scipy.dot(inv(M.T), Mc.T)  # (this works only for nbf == n)
```
[the remark this answers is not in the document]
Comment (MikhailRyazanov) [reply abel/basex.py]: See https://github.com/PyAbel/PyAbel/commit/34f6fb620975a343c52dd6fbcad4eefe5c6dffee.
@DanHickstein, is it important?

File: abel/basex.py [hunk 7]
```diff
@@ -402,33 +421,19 @@ def _bs_basex(n_vert=1001, n_horz=501,
             # End of vectorized third loop
             aux *= 2
 
-            M_horz[Rm_h - 1 + l, k] = aux  # All rows below center
-            M_horz[Rm_h - 1 - l, k] = aux  # All rows above center
+            M[Rm_h - 1 + l, k] = aux  # All rows below center
+            # M[Rm_h - 1 - l, k] = aux  # All rows above center
```
[the remark this answers is not in the document]
Comment (MikhailRyazanov) [reply abel/basex.py]: Which?

This method must be refactored, but I did not understand yet how it works. Partially, because there is no human-readable description. If these are removed, it will become even more cryptic.

File: abel/basex.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rth) [reply abel/basex.py]: Not too important, and for linalg there may indeed be a difference but scipy does not implement the dot product, [it's a numpy function](https://github.com/scipy/scipy/blob/016a6ef37f68ebdb9a2a0901f2a1131562441acc/scipy/__init__.py#L71),
```py
>>> id(numpy.dot) == id(scipy.dot)
True
```
so there is also no real reason to import it from scipy.

File: abel/basex.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rth) [reply abel/basex.py]: I meant that in
```py
            M[Rm_h - 1 + l, k] = aux  # All rows below center
            # M[Rm_h - 1 - l, k] = aux  # All rows above center
```
the second line is a copy paste of the fist, but commented out, and could be removed..

[the remark this answers is not in the document]
Comment (MikhailRyazanov) [reply abel/basex.py]: It is not. Notice the sign difference:
<pre>
           ↓
M[Rm_h - 1 <b>+ l</b>, k] = aux  # All rows <b>below</b> center
M[Rm_h - 1 <b>- l</b>, k] = aux  # All rows <b>above</b> center
</pre>

File: abel/basex.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MikhailRyazanov) [reply abel/basex.py]: Seems to be true. At least [scipy.linalg docs](https://docs.scipy.org/doc/scipy/reference/tutorial/linalg.html) use `A.dot(B)` instead of `scipy.dot(A, B)`. Let's see what @DanHickstein can say.

If we are going to convert `.dot` back `ndarray`'s method, do you think whether
`R = M.dot(inv((M.T).dot(M) + E))`
is more readable than
`R = M.dot(inv(M.T.dot(M) + E))`
or not?

(Conversion to `numpy.matmul` or `@`, as [`numpy.dot` description](https://docs.scipy.org/doc/numpy/reference/generated/numpy.dot.html) suggests is probably not a good option yet.)

**Comment by stggh** [comment]
I vote this PR #227 (and @rth's PR #228) be merged. The trivial stuff may be fixed later.

The same basis caching/storing techniques are applicable to the `dash` methods. In particular, @MikhailRyazanov's [neat idea](https://github.com/PyAbel/PyAbel/issues/225#issuecomment-422189131) of slicing any larger side pre-calculated basis.

File: abel/basex.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DanHickstein) [reply abel/basex.py]: At one point I was switching all of the linear algebra to scipy, since I thought that it was faster than numpy in the case where the user didn't have fast (Apple Accelerate Framework or Intel MKL) BLAS/LAPACK libraries. I didn't realize that the dot function was actually in numpy. Fine with me to swith it to numpy.

[the remark this answers is not in the document]
Comment (DanHickstein) [reply abel/basex.py]: I think that @rth is just wanting `scipy.dot` to be changed to `np.dot`, and not necessarily to use the `M.dot(X)` syntax. I think that both are equally clear, but that it does make some sense to use `np.dot` if `scipy.dot` simply redirects to `np.dot` anyway.

**Comment by DanHickstein** [comment]
Is this ready to merge? Looks good as far as I can tell!

**Comment by stggh** [comment]
Yes, ASAP. I have the `dasch` cached basis changes (just about) ready to submit.

**Comment by MikhailRyazanov** [comment]
> Is this ready to merge?

From my perspective, yes.

Now I also have basis cropping, memory-cache cleanup and faster basis generation, but they will go as a separate patch. (I also want to see what @stggh has done for `dasch` caching to be consistent.)

**Comment by DanHickstein** [comment]
Okay, merging! 

HUGE thanks to @MikhailRyazanov for these changes. Our BASEX implementation is now fundamentally better.
